sugondat/chain/pallets/length-fee-adjustment/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+/// using `FeeMultiplierUpdate`, this associated type is called within the `on_finalize`
+/// function of the `transaction_payment` pallet, with the purpose of converting the existing
+/// `targeted_fee_adjustment` to a new one based on network congestion.
+///
+/// The goal of this pallet is to achieve a modified fee calculation formula:
+///
+/// ```ignore
+/// inclusion_fee = base_fee + [targeted_length_fee_adjustment * length_fee] + [targeted_weight_fee_adjustment * weight_fee];
+/// ```
+///
+/// `targeted_fee_adjustment` becomes `targeted_weight_fee_adjustment`,
+/// while the behavior remains the same.
+/// `targeted_length_fee_adjustment` is a new multiplier associate to `length_fee`.
+/// This formula is achievable because the `transaction_payment`
+/// pallet uses the `compute_fee_raw` function, which computes the final fee associated with an
+/// extrinsic. This function utilizes the associated type `LengthToFee`, which converts the length
+/// of an extrinsic to a fee.
+///
+/// By default, the implementation of `LengthToFee` is a constant multiplication. However, we
+/// aim to achieve a dynamic formula, thanks to the new multiplier, implementing `sp_weights::WeightToFee`
+/// for the Pallet struct and thus being able to use it as value for the associated type `LengthToFee` in the
+/// `pallet_transaction_type::Config`. The same pallet implements aslo `MultiplierUpdate` so it can be also
+/// used for the  `FeeMultiplierUpdate` associated type.
+///
+/// `targeted_length_fee_adjustment` is updated inside `on_finalize` while
+/// `targeted_weight_fee_adjustment` is updated by the method `Convert` called by the pallet `transaction_payment`.
+#[frame_support::pallet]
+pub mod pallet {
+
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+    use pallet_transaction_payment::{
+        Multiplier, MultiplierUpdate, OnChargeTransaction, TargetedFeeAdjustment,
+    };
+    use sp_runtime::{
+        traits::{Convert, Get},
+        FixedPointNumber, Perquintill, SaturatedConversion, Saturating,
+    };
+
+    /// Configure the pallet by specifying the parameters and types on which it depends.
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_transaction_payment::Config {
+        // `targeted_weight_fee_adjustment` parameters
+        #[pallet::constant]
+        type TargetBlockFullness: Get<Perquintill>;
+        #[pallet::constant]
+        type AdjustmentVariableBlockFullness: Get<Multiplier>;
+        #[pallet::constant]
+        type MinimumMultiplierBlockFullness: Get<Multiplier>;
+        #[pallet::constant]
+        type MaximumMultiplierBlockFullness: Get<Multiplier>;
+
+        // `targeted_length_fee_adjustment` parameters
+        #[pallet::constant]
+        type TransactionByteFee: Get<<<Self as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<Self>>::Balance>;
+        #[pallet::constant]
+        type MaximumBlockLength: Get<u32>;
```

Comment:
With this being a pallet, it's better for it to be completely decoupled from the `TargetedFeeAdjustment` logic.

Reply:
There is no benefit, from what I can see, to a user configuring `pallet_transaction_payment` to use this as a multiplier instead of just providing `TargetedFeeAdjustment` before.

I did originally suggest this pallet to implement `Multiplier`, but it's actually a bad idea and I believe we decided not to do it.